Checkout/Sample/Source/3DS/WebViewController.swift

```diff
@@ -0,0 +1,90 @@
+//
+//  WebViewController.swift
+//
+//
+//  Created by Harry Brown on 01/03/2022.
+//
+
+import UIKit
+import WebKit
+import Checkout
+
+final class WebViewController: UIViewController {
+  struct ViewModel {
+    let onDismiss: ((_ webViewResult: Result<String, ThreeDSError>) -> Void)?
+    let webViewConfig: WebViewConfig
+
+    enum WebViewConfig {
+      case url(challengeURL: URL, successURL: URL, failureURL: URL)
+      case demo
+    }
+  }
+
+  var viewModel: ViewModel? {
+    didSet {
+      guard let viewModel = viewModel else {
+        return
+      }
+
+      switch viewModel.webViewConfig {
+      case .demo:
+        threeDSWKNavigationHelper = ThreeDSWKNavigationHelper(
+          successURL: URL(string: "https://www.checkout.com/"),
+          failureURL: URL(string: "https://www.example.com/")
+        )
+        threeDSWKNavigationHelper?.delegate = self
+        webView.navigationDelegate = threeDSWKNavigationHelper
+        webView.loadHTMLString(Self.demoHTML, baseURL: nil)
+
+      case let .url(challengeURL, successURL, failureURL):
+        threeDSWKNavigationHelper = ThreeDSWKNavigationHelper(successURL: successURL, failureURL: failureURL)
+        threeDSWKNavigationHelper?.delegate = self
+        webView.navigationDelegate = threeDSWKNavigationHelper
+        webView.load(URLRequest(url: challengeURL))
+      }
+    }
+  }
+
+  private let webView = WKWebView()
+  private var threeDSWKNavigationHelper: ThreeDSWKNavigationHelper?
+
+  override func loadView() {
+    view = webView
+  }
+
+  private static let demoHTML = """
+  <html>
+    <style>
+      div {
+        display: flex;
+        flex-direction: column;
+        height: 80%;
+        justify-content: space-evenly;
+        align-items: center;
+      }
+
+      a {
+        font-size: 3em;
+      }
+
+      h1 {
+        font-size: 5em;
+      }
+    </style>
+    <div>
+      <h1>3DS demo challenge</h1>
+      <a href="https://www.checkout.com/?cko-payment-token=test-token">Success</a>
+      <a href="https://www.checkout.com/">Success without token</a>
+      <a href="https://www.example.com/">Failure</a>
+      <a href="https://www.google.com/">Other</a>
+    </div>
+  </html>
+  """.trimmingCharacters(in: .whitespaces)
+}
+
+extension WebViewController: ThreeDSWKNavigationHelperDelegate {
+  func threeDSWKNavigationHelperDelegate(didReceiveResult result: Result<String, ThreeDSError>) {
+    navigationController?.popViewController(animated: true)
```

Comment:
Is there any chance this could be called from a background thread? Given the flow is relying on redirects, I think those redirects might be dispatched on threads other than main at some point, leading to the whole chain causing runtime warnings.

Personally I prefer enforcing main thread at the UI layer, when data is presented (here), but I have worked in teams where they'd want to enforce it at origin of the data (web view delegate updates?).